File: RoadRunnerJS/src/jsTest/kotlin/PathTest.kt
import TestUtil.assertDerivEquals
import path.Path
import path.PathSegment
import path.QuinticSpline
import util.DoubleProgression
import kotlin.test.Test
import kotlin.test.assertTrue

class PathTest {
    @Test
    fun testPathDerivatives() {
        val splineSegment = Path(
            listOf(
                PathSegment(
                    QuinticSpline(
                        QuinticSpline.Knot(0.0, 0.0, 20.0, 40.0),
                        QuinticSpline.Knot(45.0, 35.0, 60.0, 10.0)
                    )
                )
            ).toTypedArray()
        )

        val resolution = 1000
        val s = DoubleProgression.fromClosedInterval(0.0, splineSegment.length(), resolution)
        val ds = s.step

        val x = s.map { splineSegment[it].x }
        val dx = s.map { splineSegment.deriv(it).x }
        val d2x = s.map { splineSegment.secondDeriv(it).x }

        val y = s.map { splineSegment[it].y }
        val dy = s.map { splineSegment.deriv(it).y }
        val d2y = s.map { splineSegment.secondDeriv(it).y }

        val heading = s.map { splineSegment[it].heading }
        val headingDeriv = s.map { splineSegment.deriv(it).heading }
        val headingSecondDeriv = s.map { splineSegment.secondDeriv(it).heading }

        assertTrue(assertDerivEquals(x, dx, ds, 0.01))
        assertTrue(assertDerivEquals(dx, d2x, ds, 0.01))

        assertTrue(assertDerivEquals(y, dy, ds, 0.01))
        assertTrue(assertDerivEquals(dy, d2y, ds, 0.01))

        assertTrue(assertDerivEquals(heading, headingDeriv, ds, 0.01))
        assertTrue(assertDerivEquals(headingDeriv, headingSecondDeriv, ds, 0.01))
    }
}

File: src/roadrunnerjs/path/Path.test.ts
import { Knot, QuinticSpline } from "./QuinticSpline";
import { Path } from "./Path";
import { PathSegment } from "./PathSegment";
import { DoubleProgression, expectDerivEquals } from "roadrunnerjs/util";

test("Path Derivatives", () => {
  const splineSegment = new Path(
    new PathSegment(
      new QuinticSpline(
        new Knot({ x: 0.0, y: 0.0, dx: 20.0, dy: 40.0 }),
        new Knot({ x: 45.0, y: 35.0, dx: 60.0, dy: 10.0 })
      )
    )
  );

  const resolution = 1000;
  const s = DoubleProgression.fromClosedInterval(
    0.0,
    splineSegment.length(),
    resolution
  );
  const ds = s.step;

  const x = s.map((e) => splineSegment.get(e).x);
  const dx = s.map((e) => splineSegment.deriv(e).x);
  const d2x = s.map((e) => splineSegment.secondDeriv(e).x);

  const y = s.map((e) => splineSegment.get(e).y);
  const dy = s.map((e) => splineSegment.deriv(e).y);
  const d2y = s.map((e) => splineSegment.secondDeriv(e).y);

  const heading = s.map((e) => splineSegment.get(e).heading);
  const headingDeriv = s.map((e) => splineSegment.deriv(e).heading);
  const headingSecondDeriv = s.map((e) => splineSegment.secondDeriv(e).heading);

  expectDerivEquals(x, dx, ds, 0.01);
  expectDerivEquals(dx, d2x, ds, 0.01);

  expectDerivEquals(y, dy, ds, 0.01);
  expectDerivEquals(dy, d2y, ds, 0.01);

  expectDerivEquals(heading, headingDeriv, ds, 0.01);
  expectDerivEquals(headingDeriv, headingSecondDeriv, ds, 0.01);
});
